src/retrieval/visual_index.h

```diff
@@ -232,69 +235,219 @@ void VisualIndex<kDescType, kDescDim, kEmbeddingDim>::QueryWithVerification(
+
+    for (int j = 0; j < word_ids.cols(); ++j) {
+      const int word_id = word_ids(i, j);
+
+      if (word_id != InvertedIndexType::kInvalidWordId) {
+        inverted_index_.ConvertToBinaryDescriptor(word_id, descriptor,
+                                                  &query_entries[i].descriptor);
+
+        const auto idf_weight = inverted_index_.GetIDFWeight(word_id);
+        const auto squared_idf_weight = idf_weight * idf_weight;
+
+        inverted_index_.FindMatches(word_id, image_ids, &word_matches);
+
+        for (const auto& match : word_matches) {
+          const size_t hamming_dist =
+              (query_entries[i].descriptor ^ match->descriptor).count();
+
+          if (hamming_dist <= hamming_dist_weight_functor.kMaxHammingDistance) {
+            const float dist =
+                hamming_dist_weight_functor(hamming_dist) * squared_idf_weight;
+
+            auto& feature_matches = image_matches[match->image_id];
+            const auto feature_match = feature_matches.find(match->feature_idx);
+
+            if (feature_match == feature_matches.end() ||
+                feature_match->first < dist) {
+              feature_matches[match->feature_idx] = std::make_pair(dist, match);
+            }
+          }
+        }
+      }
+    }
+
+    // Finally, cross-reference the query and db feature matches.
+    for (const auto& feature_matches : image_matches) {
+      const auto image_id = feature_matches.first;
+
+      for (const auto& feature_match : feature_matches.second) {
+        const auto feature_idx = feature_match.first;
+        const auto dist = feature_match.second.first;
+        const auto db_match = feature_match.second.second;
+
+        const auto entry_pair = std::make_pair(&query_entries[i], db_match);
+
+        query_to_db_matches[image_id][i].emplace_back(dist, entry_pair);
+        db_to_query_matches[image_id][feature_idx].emplace_back(dist,
+                                                                entry_pair);
       }
     }
   }
 
   // Verify top-ranked images using the found matches.
-  for (size_t i = 0; i < num_verifications; ++i) {
-    auto& image_score = (*image_scores)[i];
-    const auto& geometry_matches = image_matches[image_score.image_id];
+  for (auto& image_score : *image_scores) {
+    auto& query_matches = query_to_db_matches[image_score.image_id];
+    auto& db_matches = db_to_query_matches[image_score.image_id];
 
     // No matches found.
-    if (geometry_matches.empty()) {
+    if (query_matches.empty()) {
       continue;
     }
 
-    // Collect matches for all features of current image.
+    // Enforce 1-to-1 matching: Build Fibonacci heaps for the query and database
+    // features, ordered by the minimum number of matches per feature. We'll
+    // select these matches one at a time. For convenience, we'll also pre-sort
+    // the matched feature lists by matching score.
+
+    typedef boost::heap::fibonacci_heap<std::pair<int, int>> FibonacciHeapType;
```

Comment:
Is this faster than ``std::priority_queue``?

Reply:
Yes, Fibonacci heaps allow for constant-time priority increases, which the [algorithm exploits] (https://github.com/trueprice/colmap/blob/76fb50fbd8ea44082a237396631c2c5add4f7519/src/retrieval/visual_index.h#L430) to update the priorities of other features once a feature is matched.

Reply:
Learned something new. Thanks.